File: backend/EcommerceApp/Library/src/main/java/com/ecommerce/library/service/ProductService.java
package com.ecommerce.library.service;

import com.ecommerce.library.dto.ProductDto;
import com.ecommerce.library.model.Product;
import org.springframework.data.domain.Page;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface ProductService {
    Product findById(Long id);
    List<Product> findAll();
    List<Product> getProductsBySubCategoryId(Long id);
    Product save(MultipartFile imageProduct, ProductDto product);
    Product update(MultipartFile imageProduct, ProductDto productDto);

    void enableById(Long id);
    void deleteById(Long id);

    List<ProductDto> products();
    List<ProductDto> allProduct();

    ProductDto getById(Long id);
    List<ProductDto> findByCategoryId(Long id);
    List<ProductDto> searchProducts(String keyword);
    int reduceProductStock(Long id, int qty);

    // for admin
    Page<ProductDto> searchProducts(int pageNo, String keyword);
    Page<ProductDto> getAllProducts(int pageNo);

    // for customer
    Page<ProductDto> getProductsByCategoryPaginated(Long categoryId, int page, int size);
    Page<ProductDto> getProductsBySubCategoryPaginated(Long subCategoryId, int page, int size);

    Page<ProductDto> getAllProductsPaged(int page, int size);
}

File: frontend/sportnest-customer-ui/src/services/productService.js
import api from "./customer-helper";

export const getAllProducts = async () => {
    const res = await api.get("/product/get/all");
    return res.data;
};

export const getProductById = async (id) => {
  const res =  await api.get(`/product/get/${id}`);
  return res;
};

export const getAllProductsPaged = async (page = 0, size = 8) => {
    const res = await api.get("/product/get/all/paged", {
        params: { page, size }
    });
    return res.data;
};

export const getProductsByCategoryPaged = async (categoryId, page = 0, size = 8) => {
    const res = await api.get("/product/get/by-category/paged", {
        params: { categoryId, page, size }
    });
    return res.data;
};

export const getProductsBySubCategoryPaged = async (subCategoryId, page = 0, size = 8) => {
    const res = await api.get("/product/get/by-subcategory/paged", {
        params: { subCategoryId, page, size }
    });
    return res.data;
};
